New API and documentation cleanup

przytu1 opened the pull request:
### Link to backlog task in asana (if there is no task pls create it!)

Before submitting a PR please answer:
* Was it necessary to update README? Is README updated?
* Are unit tests added for new logic?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/appsilon/data.validator/8)
<!-- Reviewable:end -->

krystian8207 commented on R/assertions.R:
```diff
@@ -0,0 +1,82 @@
+      skip_chain_opts = FALSE,
+      obligatory = obligatory,
+      description = format_description(description),
+      success_fun = assertr::success_append,
```
@przytu1 I'd prefer the approach initially proposed by me: https://github.com/Appsilon/data.validator/pull/8/commits/4f90f83c9d3c3f5e8f5bb7f4ea06e301f08a1a80
This implementation limits the features offered by assertr a lot. Any parameter like `skip_chain_opts` cannot be overwritten in the external one (parameters duplication error).

przytu1 replied:
Good point. I agree.

krystian8207 commented on R/assertions.R:
```diff
@@ -0,0 +1,82 @@
+      data = data, predicate_generator = predicate, ...,
+      skip_chain_opts = FALSE,
+      obligatory = obligatory,
+      description = format_description(description),
```
Should we use force user to use `format_description` here? This adds time by default to the description that is then displayed in the report and can make it bad looking.

przytu1 replied:
Explained in a comment below

krystian8207 commented on R/report.R:
```diff
@@ -1,5 +1,5 @@
-Validator <- R6::R6Class(
-  classname = "Validator",
+Report <- R6::R6Class(
```
The new class name surprised me. I think I got used to the previous one, but now I start to like `Report` :D

I'm curious why you decided to remove this: https://github.com/Appsilon/data.validator/pull/8/commits/f1c22fcf5acecbd1b425965e8b3b6e14a174330d#diff-a49a797e82cb8d4bb9dae1e29fcb4bc2e6ba5216cbe790ba38b6ef2095586a9cR6

This makes the usage of the package much easier. You don't even need to know that sth like `Report` class exist.

Are there any reasons for that?

przytu1 replied:
I am glad you like it :) What does "validator" even mean? For a begginer it can be super ambiguous that you actually don't have any validation functions thanks to "validator" but you accumulate results there. Weird. You accumulate results in the report, you can display it, etc.

Why I removed `getDefaultValidator()`: the reason is that it overcomplicates the implementation of our package. In every assert_* function we have to remember to pass this argument in the pipe chain (previously as `attr(data, "data-validator")` ) only to have invisible `add_results()` in the end. To me it is not worth the benefit of "not knowing that Report exists". In my opinion it is the opposite - a developer must exactly know that Report object exists, to build correct mental model how data.validator works (so it is not like a magic):

1. I create a Report object
2. I run validate(data) and assertion pipeline
3. I save the results to the report
4. I can do with the Report whatever I want (display it, analyse, reuse again, I can even have multiple independent reports at once)

I can imagine you can have different point of view on that. I would love to discuss it this week on a call :)

przytu1 replied:
Additional reason is that global variable like `.validator` in a package most probably will not pass CRAN checks. I think we had such issues with shiny.worker. Also I found some mentions like this: https://stackoverflow.com/questions/55604073/cran-policy-on-use-of-global-variables

krystian8207 replied:
I understand and agree. This simplifies the implementation a lot.
I'm still thinking with the global report we'd simplify the api a lot but let's leave it for now.

krystian8207 commented on R/utils.R:
```diff
@@ -21,3 +21,10 @@ generate_id <- function () {
+#' Return description with datetime prefix.
+#' @param description character
+#' @return character
+format_description <- function(description) {
```
What are the reasons to introduce such function?

przytu1 replied:
During testing I realised that if you run your validation pipeline multiple times, you will just accumulate error / success counter and it will be unclear if e.g. 6 violations mean 6 problematic entries / rows, or your run your code 2 times and had 3 violations.

But I think it is a discussion for another pull request. I am removing it for now and I would like to get back to it after we agree on the data.validator new API.

krystian8207 replied:
Let's address it later then.

krystian8207 commented on README.md:
```diff
@@ -7,67 +7,76 @@ data.validator
+
+# Installation
+
+For now we use Appsilon `assertr` fork, with custom functionalities we needed for `data.validator`. 
```
This description can be updated.

krystian8207 reviewed (changes requested):
@przytu1 could you address my comments. I'd like to understand more a few changes.

krystian8207 commented on README.md:
```diff
@@ -16,6 +16,11 @@ store in logs folder, or generate automatically with RStudio Connect.
 
 # How to validate data with assertr
 
+TODO: instalacja wersji ropensci z mastera! (trzeba uporządkować nasz fork i jego rekomendować)
```
to be removed

przytu1 replied:
This is weird, I don't see this TODO in the recent code. This seems like a comment to some development commit with temporary changes before issuing a pull request. I suggest we should merge with squash, but please let me know if you prefer keeping the commit history and removing this TODO from the history

krystian8207 replied:
My bad, I've been checking changes commit by commit that's why I saw it.

krystian8207 commented on R/results_parsers.R:
```diff
@@ -41,7 +41,7 @@ parse_errors_to_df <- function(data) {
     purrr::map_df(
       ~ tibble::tibble(
-        assertion.id = .$assertion.id,
+        assertion.id = generate_id(), # TODO(pawel): .$assertion.id,  Error: Column `assertion.id` not found in `.data`
```
this should be fixed now, do you know how to recreate the issue?

przytu1 replied:
Yes, now it should work correctly. I removed this TODO

krystian8207 replied:
I mean this should work fine with having .assertion.id. I'll double check it.

przytu1 commented on R/assertions.R:
```diff
@@ -0,0 +1,82 @@
+}
+
+#' @export
+assert_cols <- function(data, predicate, ..., predicate_calculated_from_column = FALSE, obligatory = FALSE, description = NA) {
```
Thank you so much for pushing about the naming. I just realised that the problem is because of ellipsis (...) used for passing column names to validate. This forces you to read the documentation and verify what is this ellipsis for, but moreover it complicates named/unnamed argument usage. When I removed it, I also realised that is causes funny code duplication (`assert` and `insist` are separate functions only because there is an ellipsis and we needed separate arguments `predicate` and `predicate_generator` that couldn't be used together:

These functions are in fact the same:

```
insist(data, predicate_generator, ..., success_fun = success_continue,
       error_fun = error_stop, skip_chain_opts = FALSE,
       obligatory = FALSE, defect_fun = defect_append, description = NA)

assert(data, predicate, ..., success_fun = success_continue,
       error_fun = error_stop, skip_chain_opts = FALSE,
       obligatory = FALSE, defect_fun = defect_append, description = NA)
```

Please take a look if you agree with my new approach.

krystian8207 replied:
Yes I agree but unfortunately we cannot replace ellipsis with `cols` argument. With such approach we're unable to pass multiple columns to the function what is the standard approach.
I'd suggest to:
1. Revert to using the ellipsis.
2. Validate if we don't pass names arguments other than the one offered by insist and assert.
I'll take care of it.

krystian8207 replied:
In the ideal world we could completely skip using separate predicate and predicate_generator arguments.
We'll just have to detect which predicate is doing what and then apply proper function. Let's stay with two parameters now but I'll try to propse sth soon.

przytu1 commented:
Thank you @krystian8207 for a very good code review. I addressed all your comments. Please check if now everything makes sense from the code perspective. Please don't merge if you approve - I would like to test it once again (I didn't have time during the weekend for thorough testing)
